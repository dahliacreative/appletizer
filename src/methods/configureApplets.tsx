import React from "react";
import { Applet } from "../components/Applet";
import history from "history";

type TConfiguredApplets<T> = Record<keyof T, React.FC<IRouterProps>>;

interface IRouterProps {
  [key: string]: any;
  history: history.History;
}

interface IAppletConfig {
  host: string;
  context?: {
    [key: string]: () => unknown;
  };
}

export function configureApplets<T extends Record<string, IAppletConfig>>(
  config: T
): TConfiguredApplets<T> {
  return Object.entries(config).reduce((apps, [key, app]) => {
    const context = Object.entries(app.context ?? {}).reduce(
      (ctxs, [key, ctx]) => ({
        ...ctxs,
        [key]: ctx(),
      }),
      {}
    );
    return {
      ...apps,
      [key]: ({ history }: IRouterProps) => (
        <Applet
          host={app.host}
          name={key}
          history={history}
          context={context}
        />
      ),
    };
  }, {} as TConfiguredApplets<T>);
}
