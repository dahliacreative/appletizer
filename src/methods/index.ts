export * from "./configureApplets";
export * from "./initializeApplet";
