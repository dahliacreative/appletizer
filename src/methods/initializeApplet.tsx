import React from "react";
import ReactDom from "react-dom/client";
import history from "history";

export interface IApplet {
  history: history.History;
  context?: {
    [key: string]: () => unknown;
  };
}

export const initializeApplet = (name: string, Applet: React.FC<IApplet>) => {
  (window as any)[`render${name}`] = (
    containerId: string,
    history: history.History,
    context: IApplet["context"]
  ) => {
    const root = ReactDom.createRoot(document.getElementById(containerId)!);
    root.render(<Applet history={history} context={context} />);
    (window as any)[`unmount${name}`] = root.unmount;
  };
};
