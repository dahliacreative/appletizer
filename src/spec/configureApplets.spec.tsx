import React from "react";
import { render } from "@testing-library/react";
import { configureApplets } from "../methods/configureApplets";
import { createBrowserHistory } from "history";
import fetchMock from "jest-fetch-mock";

fetchMock.enableMocks();

describe("configureApplets", () => {
  it("should return configured applets", async () => {
    (fetch as jest.Mock).mockResolvedValue({
      json: () => ({ files: { "main.js": "main.js" } }),
    });

    const Applets = configureApplets({
      One: {
        host: "http://localhost:3001",
        context: {
          context1: () => "context1",
          context2: () => "context2",
        },
      },
      Two: {
        host: "http://localhost:3002",
      },
    });

    expect(Applets).toEqual({
      One: expect.any(Function),
      Two: expect.any(Function),
    });

    const { container } = render(
      <Applets.One history={createBrowserHistory()} />
    );

    expect(container).toMatchSnapshot();
  });
});
