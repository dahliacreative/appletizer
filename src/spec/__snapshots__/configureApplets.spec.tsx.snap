// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configureApplets should return configured applets 1`] = `
<div>
  <main
    id="applet-One"
  />
</div>
`;
